### PR TITLE
expand contributing guide

### LINKED ISSUE #13: make a contributing guide
Perhaps as a series of wiki pages including:
- list of standard nomenclature
- code style guidelines 
- policy on exceptions
- policy on logging
- when arguments may be expected to be a DataFrame vs Series
- PR review policy/template if needed

### PR DESCRIPTION
**Pull request by wholmgren**
- [x] Closes #13 .
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

### CODE REVIEW

File: docs/source/contributing.rst [hunk 1]
```diff
@@ -3,6 +3,112 @@
...
+-------------
+
+Documentation must be written in
+`numpydoc format <https://numpydoc.readthedocs.io/>`_ format which is rendered
+using the `Sphinx Napoleon extension
+<https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html>`_.
+
+To build the docs locally, install the ``test`` dependencies specified in the
+`setup.py <https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/setup.py>`_
+file.
```
Comment (alorenzo175) [inline docs/source/contributing.rst]: Why is pytest required? Should this be docs/requirements.txt?

File: docs/source/contributing.rst [hunk 2]
```diff
@@ -3,6 +3,112 @@
...
+
+Testing
+-------
+
+The vast majority of new or modified code requires contributors to write
+new tests.
+
+A pull request will automatically run the tests for you on a variety of
+platforms (Linux, Mac, Windows) and python versions. However, it is
```
Comment (alorenzo175) [inline docs/source/contributing.rst]: Only runs on linux w/ python 3.7

File: docs/source/contributing.rst (hunk 1 above)
Comment (wholmgren) [reply docs/source/contributing.rst]: copy pasta problem from pvlib guide.

File: docs/source/contributing.rst (hunk 2 above)
Comment (wholmgren) [reply docs/source/contributing.rst]: more of the above

**Comment by wholmgren** [comment]
@cwhanse @dplarson @awig do you have any comments on this contributing guide?

**Comment by dplarson** [comment]
@wholmgren Looks good to me, both in terms of topics and content.

**Comment by cwhanse** [comment]
The table looks great!  

Is there a conda badge to go with the pypi badge?

How should I interpret the downloads?  Since some date?  Wondering if it is possible and useful to add dates to the Latest Release, and a "Downloads since" date.

File: docs/source/contributing.rst [hunk 3]
```diff
@@ -3,6 +3,112 @@
 Contributing
 ============
 
-Work in progress. In the mean time, most of the concepts in the
-`pvlib python Contributing Guide <https://pvlib-python.readthedocs.io/en/latest/contributing.html>`_
-apply here.
\ No newline at end of file
+We welcome and encourage community contributions to the Solar Forecast
```
Comment (cwhanse) [inline docs/source/contributing.rst]: "your" rather than "community", or just drop "community"

File: docs/source/contributing.rst [hunk 4]
```diff
@@ -3,6 +3,112 @@
...
+is 79 characters.
+
+Code must be compatible with Python 3.7 and above.
+
+Set your editor to strip extra whitespace from line endings. This
+prevents the git commit history from becoming cluttered with whitespace
+changes.
+
+The majority of the solarforecastarbiter-core code consists of
+functions that accept primatives or pandas.Series. Some higher-level
```
Comment (cwhanse) [inline docs/source/contributing.rst]: "primitives"

File: docs/source/contributing.rst [hunk 5]
```diff
@@ -3,6 +3,112 @@
...
+Logging
+-------
+
+Logging is encouraged in higher-level functions.
+
+DataFrames and Series
+---------------------
+
+The solarforecastarbiter-core repository focuses on analyses of time
+series data. Therefore, it is natural for the code to expect and return
```
Comment (cwhanse) [inline docs/source/contributing.rst]: "The code makes extensive use of pandas objects with DatetimeIndexes"

File: docs/source/contributing.rst [hunk 6]
```diff
@@ -3,6 +3,112 @@
...
+pandas objects with DatetimeIndexes. In most cases, code should expect
+and return one or more pandas.Series rather than a single
```
Comment (cwhanse) [inline docs/source/contributing.rst]: delete "single"

File: docs/source/contributing.rst [hunk 7]
```diff
@@ -3,6 +3,112 @@
...
+pandas.DataFrame. A function that requires multiple Series is
+self-documenting by virtue of the names of the parameters. Providing or
+expecting the wrong data from such a function raises an exception at the
+API level, where it belongs. A function that expects a single DataFrame
+must carefully document the required keys, providing the wrong data will
+result in a problem deep within the function, and expecting the wrong
+data will result in a problem far away from the function.
```
Comment (cwhanse) [inline docs/source/contributing.rst]: editorial: thanks for explaining the preference for Series over Dataframes, really helpful to understand the reasoning.

File: docs/source/contributing.rst [hunk 8]
```diff
@@ -3,6 +3,112 @@
...
+
+The exceptions to the "use Series instead of DataFrames" rule are:
+
+  * observation data may be packaged into a DataFrame with 'value' and
```
Comment (cwhanse) [inline docs/source/contributing.rst]: Separate issue: I have found the hard-wired generic column names to be a pain, when trying to use the arbiter code to test the validation functions. I have to rename the output of the validation functions to "value" etc. in order to use the plot functions.

File: docs/source/contributing.rst [hunk 9]
```diff
@@ -3,6 +3,112 @@
...
+Testing
+-------
+
+The vast majority of new or modified code requires contributors to write
+new tests.
+
+A pull request will automatically run the tests for you. However, it is
+typically more efficient to run and debug the tests in your own local
+environment.
+
```
Comment (cwhanse) [inline docs/source/contributing.rst]: A link here to instructions to set up and use an environment would be helpful, I think.

File: docs/source/contributing.rst (hunk 8 above)
Comment (wholmgren) [reply docs/source/contributing.rst]: Thanks for the feedback. https://github.com/SolarArbiter/solarforecastarbiter-core/issues/247

**Comment by wholmgren** [comment]
Thanks for the helpful suggestions.

**Pull request merged**